drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+	modeset_cleanup(fd);
+
+	ret = 0;
+
+out_close:
+	close(fd);
+out_return:
+	if (ret) {
+		errno = -ret;
+		fprintf(stderr, "modeset failed with error %d: %m\n", errno);
+	} else {
+		fprintf(stderr, "exiting\n");
+	}
+	return ret;
+}
+
+/*
+ * modeset_page_flip_event() stays the same.
+ */
+
+static void modeset_page_flip_event(int fd, unsigned int frame,
+				    unsigned int sec, unsigned int usec,
+				    void *data)
+{
+	struct modeset_dev *dev = data;
+
+	dev->pflip_pending = false;
+	if (!dev->cleanup)
+		modeset_draw_dev(fd, dev);
+}
+
+/*
+ * set_crtc_property() is a new function. It sets a property value to a CRTC
+ * object.
+ */
+
+static int set_crtc_property(drmModeAtomicReq *req,
+	struct drm_object *crtc, uint32_t crtc_id,
+	const char *name, uint64_t value)
+{
+	int i, prop_id = -1;
+
+	for (i = 0; i < crtc->props->count_props; i++) {
+		if (!strcmp(crtc->props_info[i]->name, name)) {
+			prop_id = crtc->props_info[i]->prop_id;
+			break;
+		}
+	}
+
+	if (prop_id < 0) {
+                fprintf(stderr, "no crtc property: %s\n", name);
+		return -EINVAL;
+	}
+
+	return drmModeAtomicAddProperty(req, crtc_id, prop_id, value);
+}
+
+/*
+ * set_connector_property() is a new function. It sets a property value to a
+ * connector object.
+ */
+
+static int set_connector_property(drmModeAtomicReq *req,
+	struct drm_object *connector, uint32_t conn_id,
+	const char *name, uint64_t value)
+{
+	int i, prop_id = -1;
+
+	for (i = 0; i < connector->props->count_props; i++) {
+		if (!strcmp(connector->props_info[i]->name, name)) {
+			prop_id = connector->props_info[i]->prop_id;
+			break;
+		}
+	}
+
+	if (prop_id < 0) {
+                fprintf(stderr, "no connector property: %s\n", name);
+		return -EINVAL;
+	}
+
+	return drmModeAtomicAddProperty(req, conn_id, prop_id, value);
+}
+
+/*
+ * set_plane_property is a new function. It sets a property value to a
+ * plane object.
+ */
+
+static int set_plane_property(drmModeAtomicReq *req,
+	struct drm_object *plane, uint32_t plane_id,
+	const char *name, uint64_t value)
+{
+        int i, prop_id = -1;
+
+        for (i = 0 ; i < plane->props->count_props ; i++) {
+                if (!strcmp(plane->props_info[i]->name, name)) {
+                        prop_id = plane->props_info[i]->prop_id;
+                        break;
+                }
+        }
+
+        if (prop_id < 0) {
+                fprintf(stderr, "no plane property: %s\n", name);
+                return -EINVAL;
+        }
+
+        return drmModeAtomicAddProperty(req, plane_id, prop_id, value);
+}
+
+/*
+ * modeset_atomic_modeset() is also new. It is used to atomically set some
+ * properties to our connector, CRTC and plane objects. This function is used to
+ * setup the modeset before we start drawing.
+ *
+ * Notice that we can't perform an atomic commit without a configured
+ * framebuffer. The intention here is not to draw (the framebuffer is
+ * zeroed), we just want to setup everything before we start to draw.
```

Comment:
This is not a good idea: you will be intentionally showing content you do not mean to show. This is how it is supposed to work with atomic:
- Pick video mode, crtc, primary plane, connector, etc.
- Create a framebuffer.
- Do an atomic `TEST_ONLY` commit. If it fails, bail out or change something and try again.
- Draw.
- Do the same atomic commit again, but this time without `TEST_ONLY` flag.

If you are going to drive multiple connectors, use them *all* while doing the above. Do not do this initial modeset committing connector by connector. Even if you are just going to turn off all other CRTCs or connectors, include them all in the same single atomic commit.

If you attempt to program CRTC-connector associations one at a time, it is very possible that you end up in an intermediate state that is invalid. That means you cannot reach your desired end state even if it is valid. With atomic API it is even more likely than with legacy API, because atomic does not automatically fix things up behind your back.